src/common/transformations/src/transformations/common_optimizations/shared_ops_optimization.cpp

```diff
@@ -122,20 +149,60 @@ bool reshapes_are_equal(const Node* lhs, const Node* rhs) {
            inputs_from_same_source_or_equal_constants(lhs, rhs);
 }
 
+bool shapeof_are_equal(const Node* lhs, const Node* rhs) {
+    auto lhs_output_et = element::i64, rhs_output_et = element::i64;
+    if (const auto shape = as_type<const v3::ShapeOf>(lhs)) {
+        lhs_output_et = shape->get_output_type();
+    } else if (!as_type<const v0::ShapeOf>(lhs)) {
+        return false;
+    }
+    if (const auto shape = as_type<const v3::ShapeOf>(rhs)) {
+        rhs_output_et = shape->get_output_type();
+    } else if (!as_type<const v0::ShapeOf>(rhs)) {
+        return false;
+    }
+    return lhs_output_et == rhs_output_et && inputs_from_same_source_or_equal_constants(lhs, rhs);
+}
+
+bool gathers_are_equal(const Node* lhs, const Node* rhs) {
+    const auto l_gather = as_type<const op::util::GatherBase>(lhs);
+    const auto r_gather = as_type<const op::util::GatherBase>(rhs);
+    if (!l_gather || !r_gather)
```

Comment:
what if they are different Gathers?

Reply:
The highest version will be picked to stay. You can look it up -- everything is stored in the GatherBased now.

Reply:
But the logic works this way:
1. the root op selected
2. the root op compared with each node from the list (or the sorted list in the current version) 
3. the root op and the selected one merged (the selected replaced with root)

It looks like for this case 
Relu -> (multi out: Gather v5, Gather v5, Gather v1) 
the result will be
Relu -> Gather v5 -> (multi out)

The question is: should we merge Gather v5 and Gather v1 here?

Reply:
I changed this logic. There are no base classes in this code now.